src/compiler/checker.ts

```diff
@@ -8016,6 +8154,65 @@ module ts {
             }
         }
 
+        function getAwaitedType(type: Type, fallbackType?: Type, seen: boolean[] = []): Type {
+            var seen: boolean[] = [];
+            if (!type || !globalIPromiseType || !(type.flags & TypeFlags.ObjectType)) {
+                return fallbackType;
+            }
+            seen[type.id] = true;
+            if (checkTypeRelatedTo(type, globalIPromiseType, assignableRelation, undefined)) {
+                var thenProp = getPropertyOfType(type, "then");
+                var thenType = getTypeOfSymbol(thenProp);
+                var thenSignatures = getSignaturesOfType(thenType, SignatureKind.Call);
+                var onFulfilledParameterType = getUnionType(map(thenSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));
+                var onFulfilledParameterSignatures = getSignaturesOfType(onFulfilledParameterType, SignatureKind.Call);
+                var awaitedType = getUnionType(map(onFulfilledParameterSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));

```

Comment:
I’ve changed this to be a bit cleaner and only union the types once.

From: Jason Freeman [mailto:notifications@github.com]
Sent: Friday, February 13, 2015 5:12 PM
To: Microsoft/TypeScript
Cc: Ron Buckton
Subject: Re: [TypeScript] Async Functions (#1781)


In src/compiler/checker.ts<https://github.com/Microsoft/TypeScript/pull/1781#discussion_r24709823>:

> @@ -8016,6 +8154,65 @@ module ts {

>              }

>          }

>

> +        function getAwaitedType(type: Type, fallbackType?: Type, seen: boolean[] = []): Type {

> +            var seen: boolean[] = [];

> +            if (!type || !globalIPromiseType || !(type.flags & TypeFlags.ObjectType)) {

> +                return fallbackType;

> +            }

> +            seen[type.id] = true;

> +            if (checkTypeRelatedTo(type, globalIPromiseType, assignableRelation, undefined)) {

> +                var thenProp = getPropertyOfType(type, "then");

> +                var thenType = getTypeOfSymbol(thenProp);

> +                var thenSignatures = getSignaturesOfType(thenType, SignatureKind.Call);

> +                var onFulfilledParameterType = getUnionType(map(thenSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));

> +                var onFulfilledParameterSignatures = getSignaturesOfType(onFulfilledParameterType, SignatureKind.Call);

> +                var awaitedType = getUnionType(map(onFulfilledParameterSignatures, signature => getTypeAtPosition(getErasedSignature(signature), 0)));

This seems pretty reasonable, but beware that signatures degenerate under union if they are not exactly the same (modulo return type).

—
Reply to this email directly or view it on GitHub<https://github.com/Microsoft/TypeScript/pull/1781/files#r24709823>.